symengine/tests/test_lambdify.py

```diff
@@ -0,0 +1,393 @@
+    inp = [7]
+    check(l(inp, use_numpy=use_numpy), inp)
+
+
+def test_2dim_Matrix():
+    _test_2dim_Matrix(False)
+
+
+@pytest.mark.skipif(not HAVE_NUMPY, reason='requires numpy')
+def test_2dim_Matrix_numpy():
+    if not HAVE_NUMPY:  # nosetests work-around
+        return
+    _test_2dim_Matrix(True)
+
+
+def _test_2dim_Matrix_broadcast(use_numpy):
+    l, check = _get_1_to_2by3_matrix()
+    inp = range(1, 5)
+    out = l(inp, use_numpy=use_numpy)
+    for i in range(len(inp)):
+        check(out[i, ...], (inp[i],))
+
+
+def test_2dim_Matrix_broadcast():
+    _test_2dim_Matrix_broadcast(False)
+
+
+@pytest.mark.skipif(not HAVE_NUMPY, reason='requires numpy')
+def test_2dim_Matrix_broadcast_numpy():
+    if not HAVE_NUMPY:  # nosetests work-around
+        return
+    _test_2dim_Matrix_broadcast(True)
+
+
+@pytest.mark.skipif(not HAVE_NUMPY, reason='requires numpy')
+def test_jacobian():
+    if not HAVE_NUMPY:  # nosetests work-around
+        return
+    x, y = se.symbols('x, y')
+    args = se.DenseMatrix(2, 1, [x, y])
+    v = se.DenseMatrix(2, 1, [x**3 * y, (x+1)*(y+1)])
+    jac = v.jacobian(args)
+    lmb = se.Lambdify(args, jac)
+    out = np.empty((2, 2))
+    inp = X, Y = 7, 11
+    lmb(inp, out)
+    assert np.allclose(out, [[3 * X**2 * Y, X**3],
+                             [Y + 1, X + 1]])
+
+
+@pytest.mark.skipif(not HAVE_NUMPY, reason='requires numpy')
+def test_excessive_args():
+    if not HAVE_NUMPY:  # nosetests work-around
+        return
+    x = se.symbols('x')
+    lmb = se.Lambdify([x], [-x])
+    inp = np.ones(2)
+    out = lmb(inp)
+    assert np.allclose(inp, [1, 1])
+    assert len(out) == 2  # broad casting
+    assert np.allclose(out, -1)
+
+
+@pytest.mark.skipif(not HAVE_NUMPY, reason='requires numpy')
+def test_excessive_out():
+    if not HAVE_NUMPY:  # nosetests work-around
+        return
+    x = se.symbols('x')
+    lmb = se.Lambdify([x], [-x])
+    inp = np.ones(1)
+    out = np.ones(2)
+    out = lmb(inp, out)
+    assert np.allclose(inp, [1, 1])
+    assert out.shape == (2,)
+    assert out[0] == -1
+    assert out[1] == 1
+
+
+def ravelled(A):
+    try:
+        return A.ravel()
+    except AttributeError:
+        return A
+
+
+def _get_2_to_2by2_list(real=True):
+    args = x, y = se.symbols('x y')
+    exprs = [[x + y*y, y*y], [x*y*y, se.sqrt(x)+y*y]]
+    l = se.Lambdify(args, exprs, real=real)
+
+    def check(A, inp):
+        X, Y = inp
+        assert A.shape[-2:] == (2, 2)
+        ref = [X + Y*Y, Y*Y, X*Y*Y, cmath.sqrt(X)+Y*Y]
+        ravA = ravelled(A)
+        print(dir(ravA))
+        try:
+            size = ravA.memview.size
+        except AttributeError:
+            size = len(ravA)
+        for i in range(size//4):
+            assert allclose(ravA[i*4:(i+1)*4], ref)
+    return l, check
+
+
+@pytest.mark.xfail
+def xtest_2_to_2by2_list():
+    l, check = _get_2_to_2by2_list()
+    inp = [13, 17]
+    A = l(inp, use_numpy=False)
+    check(A, inp)
```

Comment:
Good question, can't remember at the moment. I will look into it and document it properly.